### PR TITLE
Add force_for_new_devices option to renew profile if device count differs

### PR DESCRIPTION
**Pull request by NSCSKirk**
Quick background: I work on an iOS project that contains 16 targets each with their own Watch App and Watch Extension. Managing these manually across seven team members before match was truly a nightmare. Thankfully things are simplified now but there was one problem: what if someone added a new device? I could add the --force flag to every match call in my Fastfile but that seems unnecessary and would cause my repository to grow to a size thats progressively slower to clone (we use Stash so no shallow_clone for us).

Enter the --force_for_new_devices flag. The idea is match will compare the number of devices in your provisioning profile to the number of registered devices on the developer portal and, if they differ, set the --force flag. So now anytime new devices are added, as long as you have this flag passed to match, everything will take care of itself and I won't have to edit my Jenkins configuration.

Let me know what you think. I'm pretty new to Ruby so I apologize if I used methods or styles that are odd. I am open to any and all feedback!

### CODE REVIEW

**Comment by WinterKirk** [comment]
Not sure I understand the rubocop violations. Could some be false positives? Any help with this would be appreciated :)

**Comment by KrauseFx** [comment]
Yah, some of those validations are incorrect, but some are right, see https://travis-ci.org/fastlane/match/jobs/106642539. You can run the validation locally using `rubocop -a`. You can find more information on the countdown repo: https://github.com/fastlane/countdown

**Comment by KrauseFx** [comment]
I'll review the pull request over the next few days :+1:

File: lib/match/runner.rb [hunk 1]
```diff
@@ -90,5 +96,28 @@ def profile(params: nil, certificate_id: nil)
...
+        if type == "appstore"
+          profiles = Spaceship.provisioning_profile.app_store
+        elsif type == "adhoc"
+          profiles = Spaceship.provisioning_profile.ad_hoc
+        else
+          profiles = Spaceship.provisioning_profile.development
+        end
+
+        app_profiles = profiles.find_by_bundle_id(params[:app_identifier])
+        portal_profile = app_profiles.detect { |i| i.uuid == uuid }
```
Comment (KrauseFx) [inline lib/match/runner.rb]: If we already have the profile UUID we should be able to find it directly by just iterating over
```ruby
Spaceship::ProvisioningProfile.all.detect { |i| i.uuid == uuid }
```
This way it also works for enterprise profiles and other types in the future

File: lib/match/runner.rb [hunk 2]
```diff
@@ -90,5 +96,28 @@ def profile(params: nil, certificate_id: nil)
 
       return uuid
     end
+
+    def device_count_different?(params: nil, profile: nil)
```
Comment (KrauseFx) [inline lib/match/runner.rb]: Instead of passing `params`, it would be nice to pass the actual values we need, in this case probably just the app identifier.

Comment (KrauseFx) [inline lib/match/runner.rb]: By moving out the `force_for_new_devices` check and getting rid of the type detection.

File: lib/match/runner.rb [hunk 3]
```diff
@@ -74,6 +74,12 @@ def profile(params: nil, certificate_id: nil)
 
       # Install the provisioning profiles
       profile = profiles.last
+
+      if params[:force] == false
```
Comment (KrauseFx) [inline lib/match/runner.rb]: Should be 
```ruby
unless params[:force]
```

**Comment by KrauseFx** [comment]
This looks great, thank you for the pull request. Could you please update the PR according to the comments I added? Also, please make sure all tests pass locally using `rspec` and `rubocop -a`. Additionally, please squash the commits to one and let me know once it's ready :+1: Thanks again for this great PR.

**Comment by WinterKirk** [comment]
Thanks for the tips! I applied your suggestions, committed the corrections made by `rubocop -a`, updated a spec to keep it passing, and squashed the commits into one. Let me know if everything looks good. Thanks for taking the time to consider my pull request.

File: bin/match [hunk 4]
```diff
@@ -77,7 +77,7 @@ class MatchApplication
         params = FastlaneCore::Configuration.create(Match::Options.available_options, options.__hash__)
         params.load_configuration_file("Matchfile")
 
-        Match::ChangePassword.update(params: params)
+        Match::ChangePassword[:params] = params
```
Comment (KrauseFx) [inline bin/match]: Was this change intentional? I don't think that works

File: lib/match/version.rb [hunk 5]
```diff
@@ -1,4 +1,4 @@
 module Match
-  VERSION = "0.2.5"
-  DESCRIPTION = "Easily sync your certificates and profiles across your team using git"
+  VERSION = "0.2.5".freeze
```
Comment (KrauseFx) [inline lib/match/version.rb]: Please don't change this

**Comment by KrauseFx** [comment]
Oh no, I see for some reason it used the new rubocop version, that has different styling rules. We use rubocop 0.35.1. This introduced a lot of changes that shouldn't be here, I'll comment on those.

File: lib/match/version.rb [hunk 6]
```diff
@@ -1,4 +1,4 @@
...
+  DESCRIPTION = "Easily sync your certificates and profiles across your team using git".freeze
```
Comment (KrauseFx) [inline lib/match/version.rb]: Same here

File: lib/match/runner.rb [hunk 7]
```diff
@@ -90,5 +92,19 @@ def profile(params: nil, certificate_id: nil)
 
       return uuid
     end
+
+    def device_count_different?(profile: nil)
```
Comment (KrauseFx) [inline lib/match/runner.rb]: The last line of this method should return `false`

File: bin/match (hunk 4 above)
Comment (KrauseFx) [inline bin/match]: Please revert this change

**Comment by KrauseFx** [comment]
Awesome, thanks for updating the PR. I added some final comments before it's ready to be merged :+1:

**Comment by WinterKirk** [comment]
Oh man, sorry. I reverted the changes that shouldn't have been there and added the return statement and pushed again. Sorry for the confusion, I downgraded to the right version of rubocop now.

**Comment by KrauseFx** [comment]
This looks great and is ready from my side. Thanks for updating the PR. Did you test this updated change locally now?

**Comment by WinterKirk** [comment]
Yep! Tests passed with no failures and rubocop is happy with my changes.

**Comment by KrauseFx** [comment]
@WinterKirk Sorry, I meant if you tested running it for your account? You can read more about how you can run it locally here:O https://github.com/fastlane/countdown

**Comment by WinterKirk** [comment]
Oh, yeah. I used the ability to enable and disable devices in the developer portal to test it at first (which worked fine) then I added my wife's device and ran it again, which worked as I would have expected, forcing as it sensed her iPhone. Otherwise the force flag is not set. Worked in my testing.

**Comment by KrauseFx** [comment]
Perfect, thank you so much for your work! This will be in the next release :+1:
